lightning/src/ln/router.rs

```diff
@@ -952,19 +968,42 @@ impl Router {
 			if first_hops.is_none() || hop.src_node_id != network.our_node_id { // first_hop overrules last_hops
 				if network.nodes.get(&hop.src_node_id).is_some() {
 					if first_hops.is_some() {
-						if let Some(first_hop) = first_hop_targets.get(&hop.src_node_id) {
-							add_entry!(first_hop, hop.src_node_id, dummy_directional_info, 0);
+						if let Some(&(ref first_hop, ref features)) = first_hop_targets.get(&hop.src_node_id) {
+							// Currently there are no channel-context features defined, so we are a
+							// bit lazy here. In the future, we should pull them out via our
+							// ChannelManager, but there's no reason to waste the space until we
+							// need them.
+							add_entry!(first_hop, hop.src_node_id, dummy_directional_info, ChannelFeatures::with_known_relevant_init_flags(&features), 0);
 						}
 					}
-					add_entry!(hop.short_channel_id, target, hop, 0);
+					// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
+					// really sucks, cause we're gonna need that eventually.
+					add_entry!(hop.short_channel_id, target, hop, ChannelFeatures::empty(), 0);
 				}
 			}
 		}
 
 		while let Some(RouteGraphNode { pubkey, lowest_fee_to_node, .. }) = targets.pop() {
 			if pubkey == network.our_node_id {
 				let mut res = vec!(dist.remove(&network.our_node_id).unwrap().3);
-				while res.last().unwrap().pubkey != *target {
+				loop {
+					if let Some(&(_, ref features)) = first_hop_targets.get(&res.last().unwrap().pubkey) {
+						res.last_mut().unwrap().node_features = NodeFeatures::with_known_relevant_init_flags(&features);
+					} else {
+						if let Some(node) = network.nodes.get(&res.last().unwrap().pubkey) {
+							res.last_mut().unwrap().node_features = node.features.clone();
+						} else {
+							// We should be able to fill in features for everything except the last
+							// hop, if the last hop was provided via a BOLT 11 invoice (though we
+							// should be able to extend it further as BOLT 11 does have feature
+							// flags for the last hop node itself).
+							assert!(res.last().unwrap().pubkey == *target);
+						}
+					}
+					if res.last().unwrap().pubkey == *target {
+						break;
+					}
```

Comment:
Is there a reason features are set on the last hop but other fields are not? It's unclear to me.

Reply:
The RouteHop object is a little confusing - some fields are about the channel used to get to the given node and some fields are about the node itself. Every field should be set by the end of the loop, but the fee_msat and cltv_expiry_delat is set in the previous iteration.